cs.c

```diff
@@ -51,57 +51,238 @@
 
+cs_err (*cs_arch_init[MAX_ARCH])(cs_struct *) = {
 #ifdef CAPSTONE_HAS_ARM
-	ARM_enable();
+	ARM_global_init,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_ARM64
-	AArch64_enable();
+	AArch64_global_init,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_MIPS
-	Mips_enable();
+	Mips_global_init,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_X86
+	X86_global_init,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_POWERPC
-	PPC_enable();
+	PPC_global_init,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_SPARC
-	Sparc_enable();
+	Sparc_global_init,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_SYSZ
-	SystemZ_enable();
+	SystemZ_global_init,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_XCORE
+	XCore_global_init,
+#else
+	NULL,
+#endif
+};
+
+cs_err (*cs_arch_option[MAX_ARCH]) (cs_struct *, cs_opt_type, size_t value) = {
+#ifdef CAPSTONE_HAS_ARM
+	ARM_option,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_ARM64
+	AArch64_option,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_MIPS
+	Mips_option,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_X86
-	X86_enable();
+	X86_option,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_POWERPC
+	PPC_option,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_SPARC
+	Sparc_option,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_SYSZ
+	SystemZ_option,
+#else
+	NULL,
 #endif
 #ifdef CAPSTONE_HAS_XCORE
-	XCore_enable();
+	XCore_option,
+#else
+	NULL,
 #endif
+};
 
+void (*cs_arch_destroy[MAX_ARCH]) (cs_struct *) = {
+#ifdef CAPSTONE_HAS_ARM
+	ARM_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_ARM64
+	AArch64_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_MIPS
+	Mips_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_X86
+	X86_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_POWERPC
+	PPC_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_SPARC
+	Sparc_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_SYSZ
+	SystemZ_destroy,
+#else
+	NULL,
+#endif
+#ifdef CAPSTONE_HAS_XCORE
+	XCore_destroy,
+#else
+	NULL,
+#endif
+};
 
-	initialized = true;
-}
+cs_mode cs_arch_disallowed_mode_mask[MAX_ARCH] = {
+#ifdef CAPSTONE_HAS_ARM
+	~(CS_MODE_LITTLE_ENDIAN | CS_MODE_ARM | CS_MODE_V8 | CS_MODE_MCLASS | CS_MODE_THUMB | CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_ARM64
+	~(CS_MODE_LITTLE_ENDIAN | CS_MODE_ARM | CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_MIPS
+	~(CS_MODE_LITTLE_ENDIAN | CS_MODE_32 | CS_MODE_64 | CS_MODE_MICRO | CS_MODE_MIPS32R6 | CS_MODE_MIPSGP64 |
+	  CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_X86
+	~(CS_MODE_LITTLE_ENDIAN | CS_MODE_32 | CS_MODE_64 | CS_MODE_16),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_POWERPC
+	~(CS_MODE_LITTLE_ENDIAN | CS_MODE_32 | CS_MODE_64 | CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_SPARC
+	~(CS_MODE_BIG_ENDIAN | CS_MODE_V9),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_SYSZ
+	~(CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+#ifdef CAPSTONE_HAS_XCORE
+	~(CS_MODE_BIG_ENDIAN),
+#else
+	0,
+#endif
+};
 
-unsigned int all_arch = 0;
+unsigned int all_arch =
```

Comment:
this will break if ARM is not enabled, because in that case we have `all_arch = | ...`. to fix this, we can put `0` in front of it (so `all_arch = 0 | ...`)

Reply:
This actually works since the `#else` case puts in a `0`.
The bitwise OR (`|`) operators are not inside the `#ifdef ... #endif`.

My build tests seem to confirm:
~~~
$ CAPSTONE_ARCHS=powerpc make -j
  CC      cs.o
  CC      utils.o
  CC      SStream.o
  CC      MCInstrDesc.o
  CC      MCRegisterInfo.o
  CC      arch/PowerPC/PPCInstPrinter.o
  CC      arch/PowerPC/PPCDisassembler.o
  GEN     capstone.pc
  CC      arch/PowerPC/PPCMapping.o
  CC      arch/PowerPC/PPCModule.o
  CC      MCInst.o
  LINK    libcapstone.so
  AR      libcapstone.a
ar: creating ./libcapstone.a
make[1]: Entering directory 'capstone/cstool'
  CC      cstool_x86.o
  CC      cstool_arm64.o
  CC      cstool_sparc.o
  CC      cstool_ppc.o
  CC      cstool_xcore.o
  CC      cstool_arm.o
  CC      cstool_systemz.o
  CC      cstool_mips.o
  CC      cstool.o
  LINK    cstool
make[1]: Leaving directory 'capstone/cstool'
make -C tests
make[1]: Entering directory 'capstone/tests'
  CC      test_basic.o
  CC      test_detail.o
  CC      test_iter.o
  CC      test_ppc.o
  CC      test_skipdata.o
  LINK    test_basic
  LINK    test_detail
  LINK    test_skipdata
  LINK    test_iter
  LINK    test_ppc
  LINK    test_detail.static
  LINK    test_iter.static
  LINK    test_basic.static
  LINK    test_skipdata.static
  LINK    test_ppc.static
make[1]: Leaving directory 'capstone/tests'
install -m0755 ./libcapstone.so ./tests/
cd ./tests/ && mv libcapstone.so libcapstone.so.3 && ln -s libcapstone.so.3 libcapstone.so

$ ./tests/test_ppc
****************
Platform: PPC-64
Code:0x43 0x20 0x0c 0x07 0x41 0x56 0xff 0x17 0x80 0x20 0x00 0x00 0x80 0x3f 0x00 0x00 0x10 0x43 0x23 0x0e 0xd0 0x44 0x00 0x80 0x4c 0x43 0x22 0x02 0x2d 0x03 0x00 0x80 0x7c 0x43 0x20 0x14 0x7c 0x43 0x20 0x93 0x4f 0x20 0x00 0x21 0x4c 0xc8 0x00 0x21 0x40 0x82 0x00 0x14 
Disasm:
0x1000:	bdnzla+	0xc04
	op_count: 1
		operands[0].type: IMM = 0xc04
	Branch hint: 1
...
~~~

Reply:
oops you are right, i missed that 0

Reply:
or without the `#else` 
```C
all_arch = 0
#ifdef CAPSTONE_HAS_ARM
           | (1 << CS_ARCH_ARM)
#endif
#ifdef CAPSTONE_HAS_ARM64
           | (1 << CS_ARCH_ARM64)
#endif
... 
```
which saves a bit of space, not a big deal though.

Reply:
yes i would do that too, but this is also OK